# Profitable App Profiles for the App Store and Google Play Markets

## Analyzing Mobile App Data

This project is aimed towards analyzing mobile apps' data in order to understand what types of apps are more likely to attract more users. In this project, as hypothetical data analysts for an app development company, we are studying which apps are most likely to engage the most users, and therefore generate most revenue.

We work towards providing free mobile apps on both the Apple App Store and Google Play, and our main source of revenue is advertisements within those applications. By finding out which mobile apps engage the most users, we can provide our software development team with sufficient information to maximize revenue.

## Opening and Exploring the Data

As of September 2018, there approximately 2 million iOS apps available on the Apple App Store and 2.1 million Android apps on Google play. The statistic is shown in the following graph:
![Number of Mobile Apps in Various Stores](https://s3.amazonaws.com/dq-content/350/py1m8_statista.png) <p style="text-align:center"> Source: <a href="https://www.statista.com/statistics/276623/number-of-apps-available-in-leading-app-stores/" target="_blank"> Statista </a></p>

Given that collecting data from over 4 million applications requires a significant amount of time and money, we will try analyzing a smaller sample of the data instead. For that, we can use the following sources:
- [This dataset](https://www.kaggle.com/lava18/google-play-store-apps) contains data (collected in August 2018) about approximately 10,000 Android apps available on Google Play. The dataset can be downloaded directly from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv).
- [This dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) contains data (collected in July 2017) about approximately 7,000 iOS apps available on the App Store. The dataset can be downloaded directly from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv).

Let's start by opening and exploring these two data sets.

In [1]:
from csv import reader

#Starting with the Google Play data set
opened_file = open("googleplaystore.csv")
read_file = reader(opened_file)
android = list(read_file)
android_heading = android[0]
android = android[1:]

#Now the App Store data set
opened_file = open("AppleStore.csv")
read_file = reader(opened_file)
ios = list(read_file)
ios_heading = ios[0]
ios = ios[1:]

To explore the data sets more efficiently, we create a function called `explore_data()` that we can use to repeatedly print rows in a readable way.

In [2]:
def explore_data(dataset, start, end, rows_and_cols=False):
    a_slice = dataset[start:end]
    for row in a_slice:
        print(row)
        print('\n') #To seperate each row
    
    if rows_and_cols:
        print("Number of rows: ", len(dataset))
        print("Number of columns: ", len(dataset[0]))

First, let's print the headings row and the first 3 rows of the andriod apps

In [3]:
print(android_heading)
print('\n')
explore_data(android, 0, 3, rows_and_cols=True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows:  10841
Number of columns:  13


We can see that this dataset has 10,841 andriod apps, and their are 13 categories for each:

| Category | Description |
| :--- | :---: |
| "App" | App Name|
| "Category" | App Category |
| "Rating" | Average App Rating |
| "Reviews" | Total Number of Reviews |
| "Size" | App Size (in Megabytes) |
| "Installs" | Number of Installations |
| "Type" | Free or Paid |
| "Price" | Price of App in USD |
| "Content Rating" | Minimum Age Requirement |
| "Genres" | All Generes of App (including Category) |
| "Last Updated" | Date of Last App Update |
| "Current ver" | Current Version of App |
| "Andriod ver" | Minimum Required Andriod Version |

 Next, let's explore the iOS apps data set.

In [4]:
print(ios_heading)
print("\n")
explore_data(ios, 0, 3, rows_and_cols=True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows:  7197
Number of columns:  16


We can see that his data set has 7,197 iOS apps, and there are 16 categories for each:

| Column name | Description |
| :--- | :---:|
| "id" | App ID |
| "track_name" | App Name |
| "size_bytes" | App Size (in Bytes) |
| "currency" | Currency For Price Column (USD) |
| "price" | Price of App |
| "rating_count_tot" | Number of User Ratings (All Versions) |
| "rating_count_ver" | Number of User Ratings (Current Version) |
| "user_rating" | Average User Rating (All Versions) |
| "user_rating_ver" | Average User Rating (Current Version) |
| "ver" | Current App Version |
| "cont_rating" | Minimum Age Requirement |
| "prime_genre" | Primary Genre of App |
| "sup_devices.num" | Number of Devices Supported |
| "ipadSc_urls.num" | Number of Screenshots Shown for Display |
| "lang.num" | Number of Supported Languages |
| "vpp_lic" | Vpp Device Based Licensing Enabled |

## Deleting Wrong Data

The Google Play dataset has a dedicated [discussion section](), and an issue with the data has been raised in [this discussion]().
According to the discussion, entry 10472 has an incorrect rating value of 19 (maximum rating is 5). We can print the row to see that.

In [5]:
print(android[10472])
print('\n')
print(android_heading)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


We can see that the error happens because the entry is missing a value in the `Category`. To fix this issue, we will delete entry nummber 10472.

In [6]:
print(len(android)) #should print 10841
del(android[10472]) #don't run more than once
print(len(android)) #should print 10840

10841
10840


Upon inspecting the discussion section in the App Store dataset, there are no input errors. On the other hand, there are duplicates in the Google Play dataset. We can remove those for more accurate results.

## Removing Duplicate Entries

### Part One

To check for duplicate entries in the Google Play dataset, we can do the following.

In [7]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print("The number of duplicate apps in the Google Play dataset: ", len(duplicate_apps))
print('\n')
print("Some duplicate apps: ", duplicate_apps[:10])

The number of duplicate apps in the Google Play dataset:  1181


Some duplicate apps:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


Let's print some of the duplicates to see if they are exactly the same.

In [8]:
for app in android:
    name = app[0]
    if name == 'Quick PDF Scanner + OCR FREE':
        print(app)

['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80804', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


In [9]:
for app in android:
    name = app[0]
    if name == 'Google Ads':
        print(app)

['Google Ads', 'BUSINESS', '4.3', '29313', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 30, 2018', '1.12.0', '4.0.3 and up']
['Google Ads', 'BUSINESS', '4.3', '29313', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 30, 2018', '1.12.0', '4.0.3 and up']
['Google Ads', 'BUSINESS', '4.3', '29331', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 30, 2018', '1.12.0', '4.0.3 and up']


In [10]:
for app in android:
    name = app[0]
    if name == 'Slack':
        print(app)

['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


From the above data, we can see that duplicate entries have the exact same data except for the fourth position of the row (index 3), the `Reviews` column. We can infer that the data was collected at different points in time, and therefore, it is safe to assume that the entries with the highest number of reviews are the most recent ones.

Therefore, we can keep the rows with the highest number of reviews and remove the other duplicate entries. To do that we will:
- Create an empty dictionary where the key is the app name and the value is the maximum number of reviews
- Use the dictionary to update the `andriod` data set to contain only the entries with the maximum number of reviews

### Part Two

In [11]:
max_reviews = {}

for app in android:
    name = app[0]
    reviews = float(app[3]) #cast to float since the reviews are stored as strings
    if name in max_reviews and max_reviews[name] < reviews:
            max_reviews[name] = reviews
    else:
        max_reviews[name] = reviews

We saw previously that the duplicates in the Google Play dataset are 1181 apps. Therefore, the length of `max_reviews` should be the length of the `andriod` dataset minus 1181.

In [12]:
print("Length of andriod dataset without duplicates:", len(android) - 1181)
print("Length of max_reviews dictionary: ", len(max_reviews))

Length of andriod dataset without duplicates: 9659
Length of max_reviews dictionary:  9659


Next, we will use the `max_reviews` dictionary to remove the duplicates from the dataset itself.

In [13]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if max_reviews[name] == n_reviews and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
        
print("Length of cleaned dataset: ", len(android_clean)) #expected: 9659

Length of cleaned dataset:  9659


We can see that the cleaned dataset has 9,659 apps/entries as expected.

In [14]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows:  9659
Number of columns:  13


In [15]:
for app in android_clean:
    name = app[0]
    if name == 'Slack' or name == 'ZOOM Cloud Meetings' or name == 'Quick PDF Scanner + OCR FREE':
        print(app)
        print('\n')

['ZOOM Cloud Meetings', 'BUSINESS', '4.4', '31614', '37M', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 20, 2018', '4.1.28165.0716', '4.0 and up']


['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80804', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']




We can see that all duplicates have been removed successfully, and the remaining entries are the ones with the highest reviews.

We can check the App Store dataset to ensure that there are no duplicates.

In [16]:
duplicate_ios = []
unique_ios = []

for app in ios:
    app_id = app[0]
    if app_id in unique_ios:
        duplicate_ios.append(app_id)
    else:
        unique_ios.append(app_id)
        
print("Number of apps in iOS dataset: ", len(ios))
print("Number of unique apps in iOS dataset: ", len(unique_ios))
print("Number of duplicates in iOS dataset: ", len(duplicate_ios))

Number of apps in iOS dataset:  7197
Number of unique apps in iOS dataset:  7197
Number of duplicates in iOS dataset:  0


We can see that there are no duplicates in the iOS dataset.

## Removing Non-English Apps

### Part One

After further exploring the datasets, we can see that some of the apps in both datasets have non-English names and are directed at a non-English speaking audience. However, we develop apps for an English-speaking audience at our company, so we will remove said apps from the datasets to obtain more relevant and accurate results.

English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, \*, /). It is safe to assume that any app name that contains a character other than those is not an app with an English name.

The characters described above use standard ASCII encoding, which makes each character have a corresponding number between 0 and 127. The python `ord()` built-in function can check for that, and we can use it to filter out any apps with non-English characters and remove it.

In [17]:
def is_english(string):
    for char in string:
        if ord(char) > 127:
            return False
        
    return True

We can find certain apps in the dataset without english names and test the function with those apps. This can be done by manually looking through the dataset until a few apps are found.

In [18]:
print(ios[813][1])
print(ios[6731][1])
print('\n')
print(android_clean[4360][0])
print(android_clean[3091][0])
print(android_clean[180][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
Instachat 😜
Docs To Go™ Free Office Suite


In [19]:
print(is_english(ios[813][1]))
print(is_english(ios[6731][1]))
print(is_english(android_clean[4360][0]))
print(is_english(android_clean[3091][0]))
print(is_english(android_clean[180][0]))

False
False
False
False
False


We can see that some apps that have English names have 'non-English' characters, such as `Instachat 😜` and `Docs To Go™ Free Office Suite`. Losing these apps could lead to the loss of useful data in the analysis.

In [20]:
print(ord('™'))
print(ord('😜'))

8482
128540


Both characters return a value greater than 127.

### Part Two

To minimize data loss, we'll remove apps that have more than 3 ASCII characters.

In [21]:
def is_english(string):
    non_ascii = 0
    
    for char in string:
        if ord(char) > 127:
            non_ascii += 1
            
    if non_ascii > 3:
        return False
    else:
        return True
    
# name: 爱奇艺PPS -《欢乐颂2》电视剧热播, expected: False
print(is_english(ios[813][1]))

# name: 中国語 AQリスニング, expected: False
print(is_english(android_clean[4360][0]))

# name: Instachat 😜, expected: True
print(is_english(android_clean[3091][0]))

# name: Docs To Go™ Free Office Suite, expected: True
print(is_english(android_clean[180][0]))

False
False
True
True


Next, we can use the updated `is_english()` function to filter out the apps with non-English names.

In [22]:
android_eng = []
ios_eng = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_eng.append(app)
        
for app in ios:
    name = app[1]
    if is_english(name):
        ios_eng.append(app)
        
explore_data(android_eng, 0, 3, True)
print('\n')
explore_data(ios_eng, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows:  9614
Number of columns:  13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+'

After filtering out 'non-English' apps, we are left with 9614 android apps and 6183 iOS apps.

## Isolating the Free Apps

As previously mentioned, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Therefore, we should isolate the free apps from the data sets in order for our analysis to be more tailored to our requirements.

In [23]:
android_free = []
ios_free = []

for app in android_eng:
    price = app[7]
    if price == "0":
        android_free.append(app)
        
for app in ios_eng:
    price = app[4]
    if price == "0.0":
        ios_free.append(app)
        
print("Final android dataset size: ", len(android_free))
print("Final iOS dataset size: ", len(ios_free))

Final android dataset size:  8864
Final iOS dataset size:  3222


After cleaning both datasets, we have 8864 android apps remaining and 3222 iOS apps remaining.

## Most Common Apps by Genre

### Part One - Summary and Next Steps

So far we have cleaned the data by:
- Removing inaccurate data
- Removing duplicate app entries
- Removing non-English spps
- Isolating the free apps

Our goal is to use the cleaned datasets in order to determine which apps are more likely to attract the most users, therefore generating the greatest revenue.

To minimize risks and overhead, our validation strategy has three steps:
1. Build a minimal Android version of the app, and add it to Google Play
2. If the app has good response from the users, we develop it further
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Since the end goal is to add the app on both Google Play and the App Store, we need to look for app profiles that are successful in both markets. Let's begin the analysis by determining the most common genres for each market. To do this, we will build frequency tables for a few columns in our datasets.

### Part Two

We'll build two functions we can use to analyze the frequency tables:
- One function to generate frequency tables that show percentages
- Another function we can use to display the percentages in descending order

The columns that we will use those functions on are the `prime_genre` column of the App Store dataset, and the `Genres` and `Category` columns of the Google Play dataset.

In [24]:
def freq_table(dataset, index):
    n_table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in n_table:
            n_table[value] += 1
        else:
            n_table[value] = 1
    
    p_table={} # We convert the frequencies to percentages
    for key in n_table:
        percentage = (n_table[key] / total) * 100
        p_table[key] = percentage
        
    return p_table

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_to_tuple = (table[key], key)
        table_display.append(key_to_tuple)
    
    table_sorted = sorted(table_display, reverse=True)
    for entry in table_sorted:
        print(entry[1], ":", entry[0])

### Part Three

Now that we have the functions ready, we can start by analyzing the `prime_genre` column of the App Store dataset.

In [25]:
display_table(ios_free, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We can see from the table above that of the free English apps in the App Store, over 58% of them are games. Entertainment followed with \~7.88%, followed by photo & video (\~4.97%), education (\~3.66%) and social networking (\~3.29%).

From this, we can develop a general impression that the App Store (at least the part with free and English apps) is dominated by apps that are designed for fun (games, entertainment, social media, photo & video, music, sports, etc.). However, just because the apps are more common on the App store doesn't necessarily imply that these apps have the greatest number of users. Sometimes, the supply and demand of apps don't match. 

Next, we can look at the frequency table of the `Genres` and `Category` columns of the Google Play dataset.

In [26]:
display_table(android_free, 1) #Category column

FAMILY : 19.223826714801444
GAME : 9.510379061371841
TOOLS : 8.461191335740072
BUSINESS : 4.580324909747293
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.5424187725631766
SPORTS : 3.4183212996389893
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2490974729241873
HEALTH_AND_FITNESS : 3.068592057761733
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.782490974729242
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.128158844765343
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
ENTERTAINMENT : 0.8799638989169676
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 0

The Google Play generes are quire different than the App Store ones. Less apps on Google Play are designed for fun and more are aimed towards practical purposes (family, tools, business, lifestyle, productivity, etc.) relative to the App Store. However, after further investigation we found that the `Family` category (which accounts for the most apps on Google Play: more than 19%) represents games that are aimed towards younger children. 

With that being said, while games are still the most common apps, it can still be seen that the categories are more versatile (greater representation) on Google Play. Next, we can look at the Genres column:

In [27]:
display_table(android_free, 9) #Genres column

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.580324909747293
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.5424187725631766
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2490974729241873
Action : 3.1024368231046933
Health & Fitness : 3.068592057761733
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.861462093862816
Video Players & Editors : 1.782490974729242
Casual : 1.7486462093862816
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.925090252707581

The difference between the `Genres` and the `Category` columns is not crystal clear, but one thing we can notice is that the `Genres` column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the `Category` column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular is to calculate the average number of installs for each app genre. For the Google Play dataset, we can find this information in the `Installs` column. However, we do not have an equivalent column in the App Store dataset. As a workaround, we can use the `rating_count_tot` column to get the total number of user ratings as a proxy for the number of users.

Below, we calculate the average number of user ratings per app genre on the App Store:

In [28]:
genre_dict = freq_table(ios_free, 11)
users_dict = {}

for genre in genre_dict:
    total = 0
    len_genre = 0
    for app in ios_free:
        genre_app = app[11]
        if genre == genre_app:
            app_rating = float(app[5])
            total += app_rating
            len_genre += 1
    avg_rating_num = total / len_genre
    users_dict[genre] = avg_rating_num
    print(genre, ":", avg_rating_num)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


In [29]:
print("Genre with the maximum number of users:")
print(max(users_dict, key=users_dict.get), ":", max(users_dict.values()))

Genre with the maximum number of users:
Navigation : 86090.33333333333


On average, we can see that the `Navigation` genre has the most user reviews on average (\~86,000) followed by `Reference` (\~74,000). The number of reviews of the navigation apps is heavily influenced by Google Maps and Waze, which have over half a million reviews on the app store combined.

In [30]:
for app in ios_free:
    if app[11] == 'Navigation':
        print(app[1], ':', app[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


This pattern is also seen in the `Social Networking` and `Music` genres, as a few giants in the industry have a lot of reviews, skewing the results.

Our aim is to find the most popular genres, but genres like `Social Networking`, `Music`, and `Navigation` might seem more popular than they are due to the presence of a few giants in those genres. Most apps struggle to get past the 10,000 user reviews threshold, but those giants lead to the average ratings being skewed in the favour of their genres, which could be misleading.

One way to solve this problem is to remove the giants from the analysis, and then compute the averages again. However, we will leave that level of detailed analysis for later and stick to a big picture view for now.

Let's look at the `Reference` genre (as it has the second highest average) and analyze what we find.

In [31]:
for app in ios_free:
    if app[11] == 'Reference':
        print(app[1], ':', app[5]) # print name and number of ratings

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


We can see that the average number is likely skewed by apps like 'Bible' and 'Dictionary.com'. However, while these apps have a lot of reviews, the genre still has several other apps with a high review count (greater than the 10,000 review threshold); 6 to be exact. This is different compared to the navigation or social media apps, as those only have a lot of reviews for the widely-known apps only.

Given all of that, it seems that an app that fits this reference genre might be a good idea to go with. With that in mind, it is important to note how apps that are aimed towards entertainment also have a lot of users. Therefore, an idea that would encompass both of these genres would be a book app that has features to make it stand out and be more engaging that another app.

The data shows that even though games and for-fun apps are the most downloaded, they don't have the most users, which represents saturation in the market when it comes to for-fun apps. Therefore, the book app described above overcomes this problem by being different that any other entertainment app.

Other genres that would also fit our criteria include book, weather, food and drink, and finance apps. These apps aren't very prevalent, but still have a good average number of reviews without any giants skewing the results. The book app described above would overlap with the book genre, which seems to be the most beneficial given the other factors described previously.

This book app should include features that make it stand out and more engaging for users. These features could include daily quotes from the book, quizzes and trivia about the book, and a built-in dictionary so users don't have to exit the app to check the meaning of a word.

Next, let's analyze the Google Play market to see if this idea fits well there as well.

## Most Populatr Apps by Genre on Google Play

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [32]:
display_table(android_free, 5) # the Installs columns

1,000,000+ : 15.749097472924186
100,000+ : 11.563628158844766
10,000,000+ : 10.503158844765343
10,000+ : 10.209837545126353
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

Therefore, we're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

Our first step is to convert the number of installs to floats so that we can use the numbers to compute an average.

In [33]:
cat_table = freq_table(android_free, 1)
installs_dict = {}

for category in cat_table:
    total = 0
    len_category = 0
    for app in android_free:
        category_app = app[1]
        if category == category_app:
            installs_app = app[5]
            installs_app = installs_app.replace('+', '')
            installs_app = installs_app.replace(',', '')
            installs_app = float(installs_app)
            total += installs_app
            len_category += 1
    average_install_num = total / len_category
    installs_dict[category] = average_install_num
    print(category, ":", average_install_num)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1704192.3399014778
COMICS : 817657.2727272727
COMMUNICATION : 38326063.197916664
DATING : 854028.8303030303
EDUCATION : 1768500.0
ENTERTAINMENT : 9146923.076923076
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4167457.3602941176
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 12914435.883748516
FAMILY : 5180161.789906103
MEDICAL : 123064.7898089172
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 4274688.722772277
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16772838.591304347
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24790074.17721519
NEWS_AND_MAGAZINES : 

In [34]:
print("Genre with the maximum number of users:")
print(max(installs_dict, key=installs_dict.get), ":", max(installs_dict.values()))

Genre with the maximum number of users:
COMMUNICATION : 38326063.197916664


On average, communication apps have the most installs: more than 38 million installs. Let's look at this category in further detail.

In [35]:
num_big = 0
num = 0

for app in android_free:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        num_big += 1
        print(app[0], ':', app[5]) # Name then no. of installs
        
for app in android_free:
    if app[1] == 'COMMUNICATION':
        num += 1

Messenger – Text and Video Chat for Free : 1,000,000,000+
Gmail : 1,000,000,000+
imo beta free calls and text : 100,000,000+
imo free video calls and chat : 500,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
WhatsApp Messenger : 1,000,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Hangouts : 1,000,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

In [36]:
print("Number of communication apps with 100,000,000+ installs:", num_big)
print("Number of communication apps in total:", num)

Number of communication apps with 100,000,000+ installs: 27
Number of communication apps in total: 288


We can see that the data is skewed by a few apps (27 of the total 288) that have a lot of downloads, especiall apps like WhatsApp and Google Chrome, with over a billion downloads each.

If we were to remove the apps that have over 100,000,000 installs, the average installs would be reduced by approximately 10 times.

In [37]:
under_100_m = []

for app in android_free:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3593510.3486590036

We can see that the average number of installs went from approximately 38 million to 3.5 million after the big apps were removed. The runner-up category, video players, has the same issue with giants skewing the data.

In [39]:
num_big = 0
num = 0

for app in android_free:
    if app[1] == 'VIDEO_PLAYERS' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        num_big += 1
        print(app[0], ':', app[5]) # Name then no. of installs
        
for app in android_free:
    if app[1] == 'VIDEO_PLAYERS':
        num += 1

Motorola Gallery : 100,000,000+
Google Play Movies & TV : 1,000,000,000+
YouTube : 1,000,000,000+
Dubsmash : 100,000,000+
MX Player : 500,000,000+
VLC for Android : 100,000,000+
VivaVideo - Video Editor & Photo Movie : 100,000,000+
VideoShow-Video Editor, Video Maker, Beauty Camera : 100,000,000+
Motorola FM Radio : 100,000,000+


In [40]:
print("Number of video player apps with 100,000,000+ installs:", num_big)
print("Number of video player apps in total:", num)

Number of video player apps with 100,000,000+ installs: 9
Number of video player apps in total: 158


The main concern here is that these apps seem more popular than they are; the giants make the categories seem better for our goal when they are not. Penetrating a market with apps that already dominate it would be greatly difficult, and it is highly unlikely that a minimalistic app (at least at first) would stand out and attract an audience.

The game genre seems quite popular, and with it we avoid any major skewing issues. However, as previously mentioned, since the game genre has the most apps in the market and does not have the most installs/users, we can infer that the market is saturated, and it would be difficult to stand out. Therefore, we would want to avoid that market, if possible.

Since the reference genre seemed to provide an app idea with the App Store analysis, we can look at it again in the Google Play market.

In [41]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [43]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


As we can see, the only app that greatly skews the average is Google Play Books, with the rest having 100,000,000+. Since there are only a few really popular apps with not as many downloads as the other categories (showing that it is not saturated), this market shows potential. Let's see how many apps are in the middle in terms of popularity.

In [45]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, the market seems to be have a lot of libraries and book processing softwares, so our app needs to provide extra features (as mentioned in the App Store analysis section) so it can stand out.

## Conclusion

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

If the app proves to be successful, we can then expand it to include extra books that are popular and provide all the features provided for the original book for the new books.